movie.py

```diff
@@ -0,0 +1,178 @@
+import requests
+import csv
+from datetime import datetime, timedelta
+from collections import Counter
+
+
+class MovieDataPreparation:
+    def __init__(self, num_pages):
+        self.num_pages = num_pages
+        self.base_url = 'https://api.themoviedb.org/3'
+        self.headers = {
+            'accept': 'application/json',
+            'Authorization': 'Bearer eyJhbGciOiJIUzI1NiJ9.eyJhdWQiOiIzMTI3NGFmYTRlNTUyMjRjYzRlN2Q0NmNlMTNkOTZjOSIsInN1YiI6IjVkNmZhMWZmNzdjMDFmMDAxMDU5NzQ4OSIsInNjb3BlcyI6WyJhcGlfcmVhZCJdLCJ2ZXJzaW9uIjoxfQ.lbpgyXlOXwrbY0mUmP-zQpNAMCw_h-oaudAJB6Cn5c8'
+        }
+        self.movies = self.get_sample_data()
+        self.genres = self.get_genre_data()
+
+    def make_request(self, url):
+        response = requests.get(url, headers=self.headers)
+        return response.json()
+
+    def get_movie_data(self, page):
+        url = f'{self.base_url}/discover/movie?include_adult=false&include_video=false&sort_by=popularity.desc&page={page}'
+        data = self.make_request(url)
+        return data.get('results', [])
+
+    def get_genre_data(self):
+        url = f'{self.base_url}/genre/movie/list?language=en'
+        data = self.make_request(url)
+        return data.get('genres', [])
+
+    def get_sample_data(self):
+        movies = []
+        for page in range(1, self.num_pages + 1):
+            movies.extend(self.get_movie_data(page))
+        return movies
+
+    def get_all_data(self):
+        return self.movies
+
+    def get_movies_with_indexes(self):
+        movies = self.get_sample_data()
+        return movies[3:20:4]
+
+    def get_most_popular_title(self):
+        return max(self.movies, key=lambda movie: movie.get('popularity', 0)).get('title', '')
+
+    def get_movies_with_keywords(self, keywords):
+        return [movie.get('title', '') for movie in self.movies if
+                any(keyword.lower() in movie.get('overview', '').lower() for keyword in keywords)]
+
+    def get_unique_genres(self):
+        return set(genre['name'] for genre in self.genres)
+
+    def remove_movies_by_genre(self, genre):
+        genre_ids_to_remove = [g['id'] for g in self.genres if g['name'].lower() == genre.lower()]
+
+        return [movie for movie in self.movies if
+                not any(genre_id in genre_ids_to_remove for genre_id in movie.get('genre_ids', []))]
+
+    def get_popular_genre_counts(self):
+        genre_counts = Counter()
+        for movie in self.movies:
+            genre_ids = movie.get('genre_ids', [])
+            genre_counts.update(genre_ids)
+
+        popular_genres = [(genre['name'], count) for genre in self.genres for genre_id, count in genre_counts.items() if
+                          genre.get('id') == genre_id]
+        popular_genres.sort(key=lambda x: x[1], reverse=True)
+        return popular_genres
```

Comment:
you could get the most popular id from the counter and then define it's name.
you don't need all the names of genres.